### ISSUE TITLE
Losing first line item when adding line item

### ISSUE THREAD

**Issue by philmb** [JMAConsulting/biz.jmaconsulting.lineitemedit#27]
I just noticed the feature to add ad hoc line items, and was excited, since it is super-useful to me. However, when I add one or more line items, the first item details are lost, reverting to "Contribution Amount" of $0. See image, which should show line 1 as "test item" with Unit price of $20. Notice that the Total Amount still reflects the missing details.

I have replicated this behavior on 5.5.1 and 5.9.0 (Joomla 3.9)
![line-items-wrong](https://user-images.githubusercontent.com/7117661/50990906-f789b780-14d8-11e9-8dde-8246351edecc.jpeg)

**Comment by JoeMurray** [comment]
Thanks the report, @philmb , and sorry for the issue. Can you confirm you're using version 2.0 of the extension?

Also, can you confirm the following way to reproduce the issue:

1. Create a $20 contribution that is not based on a price set and that uses a financial type without taxes.
2. Edit the contribution.
3. On the edit contribution form click Add another item and add a $20 item with financial type configured with a tax of 8.5% and save the new line. At this point the unit price and total price of the first line item change from $20 to $0 (illustrated above).

@monishdeb , could you review this before tagging a new release with the work you've been doing the last few days? Thanks.

Thanks!

**Comment by philmb** [comment]
Yes, I am using v2.0.

I can report that following your alternate steps, the issue is NOT reproduced. It worked in the expected way, such as I noticed.

To outline exactly what I can do to reproduce the problem:

1. Create a new contribution.
2. Select a Financial Type for the contribution.
3. Select "Add Item(s)", and add one item (or more) for any amount with any Financial Type (sales tax makes no difference).
4. Save the contribution.
5. When I then go to edit the contribution, that item (or the first of several line items) is now named "Contribution Amount" with a unit and total price of $0. This result is also reflected if I print an invoice (without editing the contribution).

**Comment by JoeMurray** [comment]
So @monishdeb it sounds like it is a problem related to the add line item on a new unsaved contribution.

**Comment by mlutfy** [comment]
I ran into this issue as well, and to be honest I'm still not sure I fully understand the code, but I have this workaround for now:

```
diff --git a/lineitemedit.php b/lineitemedit.php
index 0012ef8..f61ee05 100644
--- a/lineitemedit.php
+++ b/lineitemedit.php
@@ -201,9 +204,31 @@ function lineitemedit_civicrm_pre($op, $entity, $entityID, &$params) {
             $params['tax_amount'] += $lineItemParams[$i]['tax_amount'];
           }
           $params['total_amount'] = $params['net_amount'] = $params['amount'] += (CRM_Utils_Array::value('line_total', $lineItemParams[$i], 0.00) + CRM_Utils_Array::value('tax_amount', $lineItemParams[$i], 0.00));
-          if (!empty($lineItemParams[$i]['line_total']) && !empty($lineItemParams[$i]['price_field_id'])) {
-            $priceSetID = CRM_Core_DAO::getFieldValue('CRM_Price_BAO_PriceField', $lineItemParams[$i]['price_field_id'], 'price_set_id');
-            if (!empty($params['line_item'][$priceSetID])) {
+
+          if (!empty($lineItemParams[$i]['line_total'])) {
+            if (!empty($lineItemParams[$i]['price_field_id'])) {
+              $priceSetID = CRM_Core_DAO::getFieldValue('CRM_Price_BAO_PriceField', $lineItemParams[$i]['price_field_id'], 'price_set_id');
+
+              if (!empty($params['line_item'][$priceSetID])) {
+                $params['line_item'][$priceSetID][$lineItemParams[$i]['price_field_id']] = $lineItemParams[$i];
+              }
+            }
+            else {
+              // Get the first key (priceSetID). Can be replaced by array_key_first in PHP 7.3
+              // https://stackoverflow.com/questions/1028668/get-first-key-in-a-possibly-associative-array
+              reset($params['line_item']);
+              $priceSetID = key($params['line_item']);
+
+              // Assumes that the PFVs have already been created (they should, done at install time)
+              // This will assign the PFV according to the line number.
+              $pfv = civicrm_api3('PriceFieldValue', 'getsingle', [
+               'name' => 'additional_item_' . ($i + 1),
+               'sequential' => 1,
+              ]);
+
+              $lineItemParams[$i]['price_field_id'] = $pfv['price_field_id'];
+              $lineItemParams[$i]['price_field_value_id'] = $pfv['price_field_value_id'];
+
               $params['line_item'][$priceSetID][$lineItemParams[$i]['price_field_id']] = $lineItemParams[$i];
             }
           }
```

if it makes sense, I can send a PR.

**Comment by philmb** [comment]
@mlutfy 's patch didn't change the behavior for me.

**Comment by monishdeb** [comment]
I believe this is fixed in latest stable version 2.3 as I was not able to replicate the issues. Heres are the use-cases I tried: 

1. Add a backoffice Contribution (w/o price-set) with $100 and a financial type without Tax  
2. Edit the contribution and add a new line-item with 10% taxable Financial type 
3. Save and visit the edit page again. 
Result: The first line-item is not changed or lost. Here's the screencast: 
![lineitem-edit](https://user-images.githubusercontent.com/3735621/87756842-44b1bc80-c827-11ea-9da4-ee6a3d06f3ed.gif)


In another use-case, I have added a contribution and added two line-items on add Contribution form. Still I wasn't able to replicate. Screencast: 
![lineitem-edit-1](https://user-images.githubusercontent.com/3735621/87757831-f43b5e80-c828-11ea-85d9-7e863f2b5745.gif)
